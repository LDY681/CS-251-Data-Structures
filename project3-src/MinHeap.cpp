#include <iostream>
#include <climits>
#include "MinHeap.h"



//implement the methods in MinHeap.h

MinHeap::MinHeap()
{
	size = 0;
	capacity = 9999;
	array = new TreeNode*[capacity];	
}

//TODO: because we dont have main, we may need this to call this function
void MinHeap::downHeapify(MinHeap *minHeap, int i) {
	int smallest = i;
	int left = 2 * i + 1;	//so we can get a tree of 0 |1 2|3 4 5 6| 7 8 9 10 11 12 13 14| and so on
	int right = 2 * i + 2;

	//handle left and right separately in case there is no right or left
	if (left < minHeap->size && minHeap->array[left]->getFrequency() < minHeap->array[smallest]->getFrequency() ) {
		smallest = left;
	}

	if (right < minHeap->size && minHeap->array[right]->getFrequency() < minHeap->array[smallest]->getFrequency() ) {
		smallest = right;
	}

	if (smallest != i) {	//if left or right of idx node is the smallest
		//swap the smallest with node of current index
		TreeNode* t = minHeap->array[smallest];
		minHeap->array[smallest] = minHeap->array[i];	//not swap the value, we need to actually swap the pointer so no *
		minHeap->array[i] = t;
		downHeapify(minHeap, smallest);	//recurse until heap is heapified
	}

}

void MinHeap::insert(TreeNode * val)
{
	//TODO:enlarge array if max reached
	if (size == capacity - 1) {
		capacity = capacity * 2;
		TreeNode **newarray = new TreeNode*[capacity];
		int j;
		for (j = 0; j < size; j++) {
			newarray[j] = array[j];
		}
		array = newarray;
	}

	this->size++;
	int i = this->size - 1;

	//up heapify
	while (i && val->getFrequency() < this->array[(i - 1) / 2]->getFrequency()) {	//when child is less than parent// when greater or equal ->end
		this->array[i] = this->array[(i - 1) / 2];
		i = (i - 1) / 2;
	}
	this->array[i] = val;
}

TreeNode * MinHeap::removeMin()
{
	if (size == 0) {
		return NULL;
	}
	TreeNode* min = this->array[0];
	this->array[0] = this->array[this->size - 1];	//replace top with last (smallest) element
	this->size--;
	downHeapify(this, 0);	//make sure last is the smallest or we heapify
	return min;
}

TreeNode * MinHeap::top() {
	return this->array[0];
}

int MinHeap::getSize()
{
	return this->size;
}

MinHeap::~MinHeap() {
	return;
}



