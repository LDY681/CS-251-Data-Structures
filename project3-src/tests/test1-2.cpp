#include "MinHeap.h"
#include <iostream>
#include <algorithm>

using namespace std;


//Test1-2: Interleaved calls to insert and removeMin
int main()
{
  srand(0);
  MinHeap * mh = new MinHeap();

  for (int i = 0; i < 50; i++)
  {
    unsigned int f = rand() % 10;
    unsigned char val = (unsigned char)('a' + f);

    TreeNode * n = new TreeNode(val, f);

    mh->insert(n);
    int s = mh->getSize();
    cout << f << " " << (int)val << " " << s << endl;

    if (i % 3 == 2)
    {
      TreeNode * m = mh->removeMin();
      if (m == NULL)
      {
        return 1;
      }

      unsigned int f = m->getFrequency();
      unsigned char val = m->getVal();
      int s = mh->getSize();
      cout << f << " " << (int)val << " " << s << endl;

    }
  }

  return 0;
}
