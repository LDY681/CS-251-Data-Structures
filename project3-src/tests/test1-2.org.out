3 100 1
6 103 2
7 104 3
3 100 2
5 102 3
3 100 4
5 102 5
3 100 4
6 103 5
2 99 6
9 106 7
2 99 6
1 98 7
2 99 8
7 104 9
1 98 8
0 97 9
9 106 10
3 100 11
0 97 10
6 103 11
0 97 12
6 103 13
0 97 12
2 99 13
6 103 14
1 98 15
1 98 14
8 105 15
7 104 16
9 106 17
2 99 16
2 99 17
0 97 18
2 99 19
0 97 18
3 100 19
7 104 20
5 102 21
2 99 20
9 106 21
2 99 22
2 99 23
2 99 22
8 105 23
9 106 24
7 104 25
2 99 24
3 100 25
6 103 26
1 98 27
1 98 26
2 99 27
9 106 28
3 100 29
2 99 28
1 98 29
9 106 30
4 101 31
1 98 30
7 104 31
8 105 32
4 101 33
2 99 32
5 102 33
0 97 34
